components/range_cache_memory_engine/src/range_manager.rs

```diff
@@ -199,127 +318,237 @@ impl RangeManager {
         }
     }
 
-    pub fn contains(&self, key: &[u8]) -> bool {
-        self.ranges.keys().any(|r| r.contains_key(key))
-    }
-
-    pub fn get_range_for_key(&self, key: &[u8]) -> Option<CacheRange> {
-        self.ranges.keys().find_map(|r| {
-            if r.contains_key(key) {
-                Some(r.clone())
-            } else {
-                None
+    pub fn get_region_for_key(&self, key: &[u8]) -> Option<Region> {
+        if let Some((key, id)) = self
+            .regions_by_range
+            .range::<[u8], (Bound<&[u8]>, Bound<&[u8]>)>((Excluded(key), Unbounded))
+            .next()
+        {
+            let meta = &self.regions[id];
+            if &meta.range.start <= key {
+                return Some(meta.region.clone());
             }
-        })
-    }
-
-    pub fn contains_range(&self, range: &CacheRange) -> bool {
-        self.ranges.keys().any(|r| r.contains_range(range))
+        }
+        None
     }
 
-    pub fn pending_ranges_in_loading_contains(&self, range: &CacheRange) -> bool {
-        self.pending_ranges_loading_data
-            .iter()
-            .any(|(r, ..)| r.contains_range(range))
+    pub fn contains_region(&self, region_id: u64) -> bool {
+        self.regions.contains_key(&region_id)
     }
 
-    fn overlap_with_range(&self, range: &CacheRange) -> bool {
-        self.ranges.keys().any(|r| r.overlaps(range))
+    pub fn check_region_status(&mut self, region: &Region) -> Option<RegionState> {
+        use RegionState::*;
+        let Some(cached_meta) = self.regions.get_mut(&region.id) else {
+            return None;
+        };
+        if cached_meta.state == Pending
+            && cached_meta.region.get_region_epoch().version != region.get_region_epoch().version
```

Comment:
Is it possible region.epoch < cached.region.epoch after some merge and split operations?